### ISSUE TITLE
Revisit consolefs

### ISSUE THREAD

**Issue by mikbras** [microsoft/openenclave#1822]
The use of consolefs needs to be revisited and perhaps better integrated with how host console standard devices are used within the enclave. Please see PR #1772 for background.

**Comment by mikbras** [comment]
Some comments from PR #1722 by Akash:

This is more of a general comment on the design of consolefs, since I'm getting confused due to conflicting things I'm seeing in the implementation:

1. Is the purpose of the consolefs to map to the host's stdin/stdout/stderr or for writing to a tty? 

- If it is the former, then you shouldn't error out on `_consolefs_lseek` and `_consolefs_getdents64`, because the host's stdin/stdout/stderr might be redirected to something that would actually cause those calls to succeed. For example, if you do `fstat(1, ...)` on a program that was executed like this `./my_program > my_file.txt`, you will see that stdout is a regular file and therefore seekable. Strangely enough, a program's standard fds can also be a directory! I think the best approach would be to have the syscall determine the appropriate error.

- If consolefs is for writing to a tty, then you should open `/dev/tty` for stdin/stdout/stderr. I'm not sure this is desirable however, because it breaks redirection and thus logging to a file. Also, if the process is running as a daemon, then there is no controlling tty, so I don't know what the expected behavior is there.

2. There are some issues with opening `/dev/[stdin/stdout/stderr]`, instead of duplicating `0,1,2`. They would have seperate entries in the kernel file description table, so they would have different file poisition, etc. This means that all the reads / writes to stdin/stdout/stderr are completely unsynchronized between the host and the guest. For example, if you have something like `./host ./enc > log.txt`, the enclave and host would overwrite the each other due to no file pointer synchronization. Further, `/dev/std...` are just symlinks to the process's 0,1,2 fds, so if the host closes or changes them before the enclave runs, the enclave would just pick up the changes. Personally, I think just doing `dup` on `0,1,2` would be a better alternative. There are some disadvantages of the dup approach though. For example, the enclave would pick up the fcntl settings and anything that changes the kernel file description object.

In my personal opinion, I think the simplest solution would be to duplicate file descriptions 0, 1, 2 to the enclave (ignore errors if you get bad fd). That would match the behavior we have today and I think that's what most people expect. You could also ocall add the ocalls to getdents and lseek for some minor edge cases. 

Eventually, we should have a discussion on how we want stdin/stdout/stderr to behave in the enclave.

**Comment by anakrish** [comment]
> The use of consolefs needs to be revisited and perhaps better integrated with how host console standard devices are used within the enclave. Please see PR #1772 for background.

CC  @dthaler @CodeMonkeyLeet @gupta-ak 
As worded, consolefs seems too SGX specific since it assumes ability to access/redirect to host standard streams. This assumption could encourage unsafe coding practices, and may not be appropriate for various enclave platforms.

**Comment by dthaler** [comment]
I agree with the summary quoted above from (I think) Akash.  I do not think that wording is SGX-specific, it applies equally to TrustZone as far as I can tell.   The internal implementation may differ, but the behavior seen by trusted applications can be the same between SGX and TZ.

**Comment by anakrish** [comment]
Does TrustZone provide a way to print to the host's standard streams and have the output be interleaved with host's own print out to the standard streams?
What is the default behavior when writing to stdout from a TrustZone? Does the output show up in the host's console output?

**Comment by dthaler** [comment]
OP-TEE does not natively provide such a way.  We implemented it in the feature.new_platforms branch via OCALLs.
